### PR TITLE
FP8 Windowed Einsums with Multiple All-Gather Dots

### PR DESCRIPTION
**Pull request by philipphack**
Enables FP8 windowed einsums with all-gathers that have multiple dot users by shifting the dequantization of the FP8 operands to the output of the while loop.

### CODE REVIEW

**Comment by reedwm** [comment]
@philipphack I don't fully understand this pass. I pinged you, asking for help to understand it. Can you check your Google Chat pings? :)

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 1]
```diff
@@ -526,35 +623,40 @@ absl::StatusOr<bool> GpuWindowedEinsumHandler::Run(
...
+  //                   /
+  //                  /
+  //                dot
   // to:
   //                       input
   //                       |
   //                       |
   //                     windowed loop
   //                       |
+  //                     DQ (FP8)
```
Comment (frgossen) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:649]: What is DQ? Dequantize?

File: xla/tests/collective_ops_test_e2e.cc [hunk 2]
```diff
@@ -781,5 +781,45 @@ ENTRY main.12 {
 
   CollectiveOpsCompareWindowedNonWindowed(kModuleReplicatedStr);
 }
+
+TEST_F(CollectiveOpsTestE2EWindowedNonWindowed,
```
Comment (frgossen) [inline xla/tests/collective_ops_test_e2e.cc:785]: The pattern depicted in the comment includes only a few ops. Can the test case be simplified?

**Review by frgossen — CHANGES_REQUESTED** [review]
Can you provide a little more context to this PR? 
- What performance improvements did you measure? 
- What patterns are you matching for? 

The test cases seem somewhat verbose to me but I might be wrong. The patterns you're matching for seem a lot simpler. Is it possible to condense the test cases to be minimal?

**Comment by abhinavgoel95** [comment]
@frgossen this PR helps enable collective matmul for FP8 gemms. This is needed to overlap Tensor Parallel communication.
The expected speedup depends on the model configs.
But in the recent MoE models, we can expect 10-20% speedup.

File: xla/service/gpu/gpu_windowed_einsum_handler.cc (hunk 1 above)
Comment (philipphack) [reply xla/service/gpu/gpu_windowed_einsum_handler.cc]: Yes. I've edited the comment.

File: xla/tests/collective_ops_test_e2e.cc (hunk 2 above)
Comment (philipphack) [reply xla/tests/collective_ops_test_e2e.cc]: I made the input HLO less complex. The FileCheck in the unit test now compares only a part of the main computation.

**Comment by philipphack** [comment]
@frgossen the patterns are FP8 windowed einsums with input dequantizations (i.e. type-conversion to a wider type like FP16 or BF16 and scaling) and all-gathers that have multiple dot users. The dequantization is moved from the all-gather of the additional dot to the output of that holds the cached all-gather result as indicated in the comment. This enables the creation of a specific FP8 GEMM custom call in gemm_rewriter.cc which replaces dequantization and dot.

(Also rebased to resolve conflicts.)

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 3]
```diff
@@ -53,11 +53,11 @@ namespace m = match;
 // into
```
Comment (frgossen) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:53]: `Enables the creation of FP8 GEMM Custom Calls for all-gather and
// reduce-scatter windowed einsums in gemm_rewriter.cc...` - Can we add a test that tests the two passes in combination?

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 4]
```diff
@@ -673,66 +638,142 @@ class WindowedEinsumVisitor : public DfsHloRewriteVisitor {
...
     // Now that we have processed all loops, we can check if there are any
     // allgather-dot pattern that we can optimize. We'd want to transform:
     //                       input
     //                       /    |
-    //                      /     |
-    //                     AG    windowed loop
-    //                     /
-    //                    /
-    //                   dot
+    //                dequantize  |
```
Comment (frgossen) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:645]: Is dequantize optional or always needed to match?

File: xla/tests/collective_ops_e2e_test.cc [hunk 5]
```diff
@@ -831,6 +831,37 @@ ENTRY main.9_spmd {
   CollectiveOpsCompareWindowedNonWindowed(kModuleReplicatedStr);
 }
 
+TEST_F(CollectiveOpsTestE2EWindowedNonWindowed,
+       WindowedEinsumE2EAllgatherMultiConsumerF8) {
+  absl::string_view kModuleReplicatedStr = R"(
+HloModule pjit__unnamed_wrapped_function_, entry_computation_layout={(f8e4m3fn[2,16,48]{2,1,0}, f8e4m3fn[48,192]{1,0}, f8e4m3fn[48,192]{1,0}, bf16[], bf16[], bf16[])->bf16[2,16,192]{2,1,0}}, allow_spmd_sharding_propagation_to_parameters={false,false,false,false}, num_partitions=4
+
+ENTRY main.12 {
```
Comment (frgossen) [inline xla/tests/collective_ops_e2e_test.cc:839]: IIUC, the pattern is `inputs --> dequant --> while loop {dynamic-slice/collective-permute/dot}` and here you are testing the dot case? 

- Is the test case minimal?
- Can we add test cases for the two other cases?

File: xla/service/gpu/gpu_windowed_einsum_handler_test.cc [hunk 6]
```diff
@@ -830,14 +830,113 @@ ENTRY main.9_spmd {
...
 
-TEST_F(GpuWindowedEinsumHanlderTest,
+TEST_F(GpuWindowedEinsumHandlerTest, AllGatherMultipleConsumersF8) {
+  constexpr absl::string_view kHloString = R"(
+HloModule pjit__unnamed_wrapped_function_, entry_computation_layout={(f8e4m3fn[2,512,24576]{2,1,0}, f8e4m3fn[24576,24576]{1,0}, f8e4m3fn[24576,24576]{1,0}, f32[], f32[], f32[])->f32[2,2048,24576]{2,1,0}}, num_partitions=4
+windowed_dot_general_body_ag {
+  param.1 = (f32[2,512,24576]{2,1,0}, f32[24576,24576]{1,0}, f32[2,2048,24576]{2,1,0}, f32[2,2048,24576]{2,1,0}, u32[]) parameter(0)
+  get-tuple-element.1 = f32[2,512,24576]{2,1,0} get-tuple-element(param.1), index=0
+  collective-permute = f32[2,512,24576]{2,1,0} collective-permute(get-tuple-element.1), channel_id=2, source_target_pairs={{0,3},{1,0},{2,1},{3,2}}
+  collective-permute.1 = f32[2,512,24576]{2,1,0} collective-permute(collective-permute), channel_id=3, source_target_pairs={{0,3},{1,0},{2,1},{3,2}}
```
Comment (frgossen) [inline xla/service/gpu/gpu_windowed_einsum_handler_test.cc:855]: This is testing the collective permute case, is it? Can we also add minimal test cases for the two remaining cases?

**Review by frgossen — CHANGES_REQUESTED** [review]

File: xla/service/gpu/gpu_windowed_einsum_handler.cc (hunk 3 above)
Comment (philipphack) [reply xla/service/gpu/gpu_windowed_einsum_handler.cc:53]: The comment was intended to indicate that there are dynamic-slice, collective-permute and dot instructions inside the all-gather and reduce-scatter loops. The instructions occur together and don't describe separate cases. I've edited the comment to make it clearer.

File: xla/service/gpu/gpu_windowed_einsum_handler_test.cc [hunk 7]
```diff
@@ -830,14 +830,113 @@ ENTRY main.9_spmd {
...
+; CHECK-DAG:         "wait_on_operation_queues":["[[OPQUEUEID1]]"],
+; CHECK-DAG:         "force_earliest_schedule":false}
 ; CHECK-NEXT:    [[CP1:[^ ]+]] = f32[2,512,24576]{2,1,0} collective-permute([[ADD5]]), channel_id=10
 ; CHECK-NEXT:  ROOT [[OUT:[^ ]+]] = (f8e4m3fn[2,2048,24576]{2,1,0}, f8e4m3fn[24576,24576]{1,0}, f32[2,512,24576]{2,1,0}, f32[2,512,24576]{2,1,0}, u32[], /*index=5*/f32[], f32[]) tuple([[GTE0]], [[GTE1]], [[ADD3]], [[CP1]], [[ADD0]], /*index=5*/[[GTE3]], [[GTE5]])
 )");
 }
 
-TEST_F(GpuWindowedEinsumHanlderTest,
+TEST_F(GpuWindowedEinsumHandlerTest, AllGatherMultipleConsumersF8) {
+  constexpr absl::string_view kHloString = R"(
```
[the remark this answers is not in the document]
Comment (philipphack) [reply xla/service/gpu/gpu_windowed_einsum_handler_test.cc:849]: I think this test is minimal. This is essentially test AgLoopsMultipleConsumersAreChained with FP8 inputs and added dequantization (type conversion and scaling).

File: xla/service/gpu/gpu_windowed_einsum_handler.cc (hunk 4 above)
Comment (philipphack) [reply xla/service/gpu/gpu_windowed_einsum_handler.cc:645]: The dequantization step is optional. When the data type is FP8 and there’s a dequantization, it will be shifted as described in the comment. This doesn't interfere with other cases.

File: xla/tests/collective_ops_e2e_test.cc (hunk 5 above)
Comment (philipphack) [reply xla/tests/collective_ops_e2e_test.cc:874]: Yes, I think the test case is minimal. The comment in gpu_windowed_einsum_handler.cc was intended to indicate that there are dynamic-slice, collective-permute and dot instructions inside the all-gather and reduce-scatter loops. The instructions occur together and don't describe separate cases. I've edited the comment to make it clearer.

**Comment by dimitar-asenov** [comment]
@frgossen @cheshire @reedwm Could you please review the latest version of this PR?

**Comment by toli-y** [comment]
>  this PR helps enable collective matmul for FP8 gemms. This is needed to overlap Tensor Parallel communication. The expected speedup depends on the model configs. But in the recent MoE models, we can expect 10-20% speedup.

@abhinavgoel95 It's great that we expect 10-20% speedup. Can you please provide exact model and hardware configuration and the speedup you are seeing with this change?

**Comment by abhinavgoel95** [comment]
@toli-y Are you able to run GPT-5B, FP8 precision, with ICI_MESH = [1,1,8], on 8 H100 GPUs?

**Comment by philipphack** [comment]
@frgossen can you PTAL?

**Comment by thomasjoerg** [comment]
@ezhulenev to nag reviewers or assign others.

File: xla/tests/collective_ops_e2e_test.cc [hunk 8]
```diff
@@ -831,6 +866,48 @@ ENTRY main.9_spmd {
...
+
+  // Verify the creation of FP8 GEMM Custom Calls on Hopper and newer
+  // architectures.
+  DebugOptions opts = GetDebugOptionsForTest();
+  opts.set_xla_gpu_threshold_for_windowed_einsum_mib(0);
+  opts.set_xla_gpu_multi_streamed_windowed_einsum(true);
+  opts.set_xla_gpu_graph_min_graph_size(200);
+  opts.set_xla_gpu_enable_triton_gemm(false);
+  opts.add_xla_disable_hlo_passes("dot-merger");
+  CollectiveOpsVerifyF8Matmul(kModuleReplicatedStr, opts);
```
Comment (frgossen) [inline xla/tests/collective_ops_e2e_test.cc:908]: Why are these options needed to test the windowed einsum handler? 
What does this have to do with the dot-merger? Why do you have to disable triton?

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 9]
```diff
@@ -94,54 +94,55 @@ absl::Status ShiftDequantizationF8(const HloComputation* comp,
...
-                                           &coll_perms[0], m::Op().Is(gte[0]))),
-                   m::Op().Is(gte[1]),
-                   m::DynamicUpdateSlice(
-                       m::DynamicUpdateSlice().WithOperand(
-                           1, m::Dot(&dots[0], m::Op().Is(gte[0]),
-                                     m::Op().Is(gte[1]))),
-                       m::Dot(&dots[1], m::Op(), m::Op().Is(gte[1])), m::Op(),
-                       m::Op(), m::Op()),
-                   m::Op(), m::Op()))) {
+  if (Match(while_root,
```
Comment (frgossen) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:115]: I don't really understand what you are matching here exactly but it does look quite specific. Does this work on a wider range of models, e.g. will this apply if I run this on maxtext Llama 2?

**Review by frgossen — CHANGES_REQUESTED** [review]
Sorry for the delay here. I still have a hard time following what this PR does and why it is an improvement as it adds quite a bit of complexity. 

Would it be possible to write down the motivation behind this change and the observed improvements? In particular, I would like to understand: 

- What the IR looks like before this pass and why that is a problem? 
- How this pass changes the IR and why that solves the problem? 
- Why this is the right place to fix the problem? 
- On what benchmarks/ workloads do you observe performance improvements? 
- Can you provide some relevant HLO that I can add to our benchmark suite to ensure this does not regress over time?

**Comment by philipphack** [comment]
@frgossen the motivation for this change is the extension of the existing functionality for collective matmuls with multiple all-gather dots in the windowed einsum handler (see [this comment](https://github.com/openxla/xla/blob/e56c9c1c51cef21cda50eddc45f2a115836a9358/xla/service/gpu/gpu_windowed_einsum_handler.cc#L676)) to support FP8 inputs. In this case, the dot operates on a dequantization of an FP8 operand, i.e. a type conversion from FP8 to a wider type followed by a multiplication by a broadcast scaling factor. The current change adds functionality for moving the dequantization from between input and all-gather to between the windowed loop and the dot as explained in [the updated comment](https://github.com/openxla/xla/blob/6adbf31a147fd8f6fea2741fd00a02e27c203cea/xla/service/gpu/gpu_windowed_einsum_handler.cc#L641). The dequantization -> dot sequence is rewritten in gemm_rewriter.cc into a specific FP8 GEMM custom call. This change does not interfere with the existing functionality of the windowed einsum handler.

The HLO of WindowedEinsumE2EAllgatherMultiConsumerF8 in collective_ops_e2e_test.cc can also be used for monitoring performance (FP8 requires Hopper or newer architectures).

File: xla/tests/collective_ops_e2e_test.cc (hunk 8 above)
Comment (philipphack) [reply xla/tests/collective_ops_e2e_test.cc:908]: The purpose of this end-to-end test is to verify the sequence of windowed einsum handler and GEMM rewriter passes. The dot merger and Triton passes are disabled since they can potentially interfere with the outcome.

File: xla/service/gpu/gpu_windowed_einsum_handler.cc (hunk 9 above)
Comment (philipphack) [reply xla/service/gpu/gpu_windowed_einsum_handler.cc:115]: This matches the structure of the while loops of all-gather or reduce-scatter windowed einsum patterns. The patterns are generated by the SPMD dot handler and should not depend on the model.

Comment (frgossen) [reply xla/service/gpu/gpu_windowed_einsum_handler.cc:115]: One example here: I find it strange that the pattern matches a specific size tuple. If you rely on the SPMD partitioner to generate this specific pattern that seems overly fragile. What if the partitioner changes? Do the transformations here depend on the number of values passed through the while loop?

File: xla/tests/collective_ops_e2e_test.cc [hunk 10]
```diff
@@ -813,6 +838,16 @@ ENTRY main.12 {
...
   CollectiveOpsCompareWindowedNonWindowed(kModuleReplicatedStr,
                                           /*disable_dot_merger=*/true);
+
+  // Verify the creation of FP8 GEMM Custom Calls on Hopper and newer
+  // architectures.
+  DebugOptions opts = GetDebugOptionsForTest();
+  opts.set_xla_gpu_threshold_for_windowed_einsum_mib(0);
+  opts.set_xla_gpu_multi_streamed_windowed_einsum(true);
+  opts.set_xla_gpu_graph_min_graph_size(200);
+  opts.set_xla_gpu_enable_triton_gemm(false);
```
Comment (frgossen) [inline xla/tests/collective_ops_e2e_test.cc:848]: How is this related to triton gemms?

File: xla/tests/collective_ops_e2e_test.cc (hunk 8 above)
Comment (frgossen) [reply xla/tests/collective_ops_e2e_test.cc:908]: Do you know why they interfer with the outcome? Is there a fundamental reason or can this be implemented more generically so that it does compose well with the rest of XLA?

**Review by frgossen — CHANGES_REQUESTED** [review]
I left a few comments but Reed will look into this in more detail. Some of the test cases still have value names based on the operations. I suspect some of these are quantization and dequantization. If you can give them meaningful names it might be easier to follow this.

File: xla/tests/collective_ops_e2e_test.cc (hunk 10 above)
Comment (philipphack) [reply xla/tests/collective_ops_e2e_test.cc:848]: The purpose of this test is to verify that gemm_rewriter.cc creates an FP8 cublasLT custom call. The Triton rewriter may attempt to alternatively rewrite this into a Triton GEMM, which would interfere with this objective.

**Comment by philipphack** [comment]
@frgossen for some reason I can't directly respond to these two comments:

> Do you know why they interfer with the outcome? Is there a fundamental reason or can this be implemented more generically so that it does compose well with the rest of XLA?

The dot merger pass has a threshold for the input dimensions which means that it generally doesn't interfere in actual models. Since the test uses relatively small inputs, the pass is manually deactivated here to test the functionality.

> One example here: I find it strange that the pattern matches a specific size tuple. If you rely on the SPMD partitioner to generate this specific pattern that seems overly fragile. What if the partitioner changes? Do the transformations here depend on the number of values passed through the while loop?

I think the general purpose of the windowed einsum handler is to transform windowed einsums that were created by the SPMD dot handler. As such, it relies on the output of that pass by design. The alternative would be the direct modification of the dot handler which is already relatively complex.

File: xla/tests/collective_ops_e2e_test.cc (hunk 10 above)
Comment (frgossen) [reply xla/tests/collective_ops_e2e_test.cc:848]: Can you add a comment?

**Review by frgossen — CHANGES_REQUESTED** [review]
> The dot merger pass has a threshold for the input dimensions which means that it generally doesn't interfere in actual models. Since the test uses relatively small inputs, the pass is manually deactivated here to test the functionality.

Can you add a comment? 

> I think the general purpose of the windowed einsum handler is to transform windowed einsums that were created by the SPMD dot handler. As such, it relies on the output of that pass by design. The alternative would be the direct modification of the dot handler which is already relatively complex.

If this pass is so tightly coupled with the dot handler, shouldn't there be tests that reflect this? Do you think it makes sense to add test cases that test the two passes together?

**Comment by philipphack** [comment]
> If this pass is so tightly coupled with the dot handler, shouldn't there be tests that reflect this? Do you think it makes sense to add test cases that test the two passes together?

@frgossen that's what the end-to-end tests in collective_ops_e2e_test.cc do. The SPMD dot handler transforms the input HLO into a collective matmul/windowed einsum which is transformed again by the windowed einsum handler.

> Can you add a comment?

Done.

**Review by reedwm — CHANGES_REQUESTED** [review]

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 11]
```diff
@@ -48,16 +48,15 @@ namespace m = match;
...
 //
-//   inputs --> while loop {dequant --> ...}.
-absl::StatusOr<bool> ShiftDequantizationF8(HloInstruction* instr) {
-  HloComputation* comp = instr->parent();
+//   inputs --> while loop {dequant --> collective-permute/dot/etc}.
+absl::Status ShiftDequantizationF8(HloComputation* comp) {
   HloInstruction* while_instr = comp->WhileCallInstruction();
-  if (!while_instr) {
-    return false;
+  if (!while_instr || while_instr->operand(0)->user_count() != 1) {
```
Comment (reedwm) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:58]: Comment why there is the check `while_instr->operand(0)->user_count() != 1`. E.g.:

The while-loop tuple operand must only be used by the while-loop, since we modify the tuple arguments

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 12]
```diff
@@ -329,28 +326,27 @@ absl::StatusOr<bool> HandleRsWindowedEinsumLoop(HloComputation* comp,
       changed = true;
     }
   }
+  // If present, move the dequantization of FP8 operands of the dot into the
+  // while loop to allow gemm_rewriter.cc to rewrite into an FP8 Custom
```
Comment (reedwm) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:330]: I should have commented before, but the triton pass, not just gemm_rewriter.cc, can rewrite into an FP8 matmul. Probably should be phrased as 

...to allow the dequantization and dot to be fused into an FP8 matmul, e.g. by gemm_rewriter.cc.

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 13]
```diff
@@ -638,12 +637,15 @@ class WindowedEinsumVisitor : public DfsHloRewriteVisitor {
...
+    // windowed einsum loop to consume the output of the loop and remove the
+    // all-gather. Now that we have processed all loops, we can check if there
+    // are any allgather-dot pattern that we can optimize. We'd want to
+    // transform:
     //                       input
     //                       /    |
-    //                dequantize  |
-    //                  (FP8)     |
+    //               dequantize   |
+    //          (optional if FP8) |
```
Comment (reedwm) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:648]: `(optional if FP8)` implies, to me at least, that the dequantize is only optional if FP8 (and otherwise required).

Change to `(optional, only if FP8)` or `(optional)` or `(only if FP8)`

File: xla/service/gpu/gpu_windowed_einsum_handler.cc [hunk 14]
```diff
@@ -48,16 +48,15 @@ namespace m = match;
...
-//   inputs --> dequant --> while loop {...}
+//   inputs --> dequant --> while loop {collective-permute/dot/etc}
 //
 // into
 //
-//   inputs --> while loop {dequant --> ...}.
-absl::StatusOr<bool> ShiftDequantizationF8(HloInstruction* instr) {
-  HloComputation* comp = instr->parent();
+//   inputs --> while loop {dequant --> collective-permute/dot/etc}.
+absl::Status ShiftDequantizationF8(HloComputation* comp) {
```
Comment (reedwm) [inline xla/service/gpu/gpu_windowed_einsum_handler.cc:56]: Document `comp` is the body computation of the while-loop.

In fact, you get a `while_body` HloComputation* local variable below, which is exactly the same as `comp`. I would rename `comp` to `while_body` (which is obvious enough it doesn't have to be documented), and then remove the definition of `while_body` below.

**Review by reedwm — CHANGES_REQUESTED** [review]

**Review by reedwm — APPROVED** [review]
